From 7db2623ecafad5d68d908a35289bb9448109a0c8 Mon Sep 17 00:00:00 2001
From: Lars Schmertmann <Lars.Schmertmann@governikus.de>
Date: Thu, 25 Sep 2025 11:45:26 +0200
Subject: a11y: Add Switch role

Distinguishing a Switch from a CheckBox is necessary on at least Android.
Windows uses the already available IToggleProvider interface, which gets
correctly recognized as a switch when used with UIA_ButtonControlTypeId.

[ChangeLog][QtGui][QAccessible] Added new Switch role that can be used
instead of CheckBox for more accurately communicating the elements role
to assistive technology.

Fixes: QTBUG-139676
Change-Id: I77d33ea5ad32e58ba01c63c418c11ccc25d19c54
Reviewed-by: Michael Weghorn <m.weghorn@posteo.de>
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry picked from commit 293032ab2f4ad7bad0ae4dc9384fbcf448784d6f)
---
 .../accessible/linux/qspiaccessiblebridge.cpp |  6 ++++
 src/gui/accessible/qaccessible.cpp            |  1 +
 src/gui/accessible/qaccessible_base.h         |  1 +
 .../android/androidjniaccessibility.cpp       |  3 +-
 .../platforms/cocoa/qcocoaaccessibility.mm    |  8 ++++-
 .../platforms/ios/quiaccessibilityelement.mm  |  9 +++++
 .../platforms/wasm/qwasmaccessibility.cpp     | 36 +++++++++++++++++++
 .../platforms/wasm/qwasmaccessibility.h       |  1 +
 .../uiautomation/qwindowsuiamainprovider.cpp  |  5 +--
 .../windows/uiautomation/qwindowsuiautils.cpp |  1 +
 10 files changed, 67 insertions(+), 4 deletions(-)

diff --git x/qtbase/src/gui/accessible/linux/qspiaccessiblebridge.cpp y/qtbase/src/gui/accessible/linux/qspiaccessiblebridge.cpp
index 1eab8846653b9a3ca87dfe76e347245932bbc45f..861c73078f6f94d6d703b05c040bc7a6fcdd5384 100644
--- x/qtbase/src/gui/accessible/linux/qspiaccessiblebridge.cpp
+++ y/qtbase/src/gui/accessible/linux/qspiaccessiblebridge.cpp
@@ -183,6 +183,12 @@ static RoleMapping map[] = {
     //: Role of an accessible object
     { QAccessible::CheckBox, ATSPI_ROLE_CHECK_BOX, QT_TRANSLATE_NOOP("QSpiAccessibleBridge", "check box") },
     //: Role of an accessible object
+#if ATSPI_ROLE_COUNT >= 132
+    { QAccessible::Switch, ATSPI_ROLE_SWITCH, QT_TRANSLATE_NOOP("QSpiAccessibleBridge", "switch") },
+#else
+    { QAccessible::Switch, ATSPI_ROLE_CHECK_BOX, QT_TRANSLATE_NOOP("QSpiAccessibleBridge", "check box") },
+#endif
+    //: Role of an accessible object
     { QAccessible::RadioButton, ATSPI_ROLE_RADIO_BUTTON, QT_TRANSLATE_NOOP("QSpiAccessibleBridge", "radio button") },
     //: Role of an accessible object
     { QAccessible::ComboBox, ATSPI_ROLE_COMBO_BOX, QT_TRANSLATE_NOOP("QSpiAccessibleBridge", "combo box") },
diff --git x/qtbase/src/gui/accessible/qaccessible.cpp y/qtbase/src/gui/accessible/qaccessible.cpp
index 510931b4760e72669972e46061dec2ea145909f8..a2fcb8a1be73f313d11db0b9de08167f8296e691 100644
--- x/qtbase/src/gui/accessible/qaccessible.cpp
+++ y/qtbase/src/gui/accessible/qaccessible.cpp
@@ -341,6 +341,7 @@ Q_STATIC_LOGGING_CATEGORY(lcAccessibilityCore, "qt.accessibility.core");
     \value Splitter         A splitter distributing available space between its child widgets.
     \value StaticText       Static text, such as labels for other widgets.
     \value StatusBar        A status bar.
+    \value [since 6.11] Switch  A switch that can be toggled on or off.
     \value Table            A table representing data in a grid of rows and columns.
     \value Terminal         A terminal or command line interface.
     \value TitleBar         The title bar caption of a window.
diff --git x/qtbase/src/gui/accessible/qaccessible_base.h y/qtbase/src/gui/accessible/qaccessible_base.h
index 27a0fd4f92efde4b3589e5a7ba39de64b7f1d317..c4697b468de9f74202dbed24d896c173f3bad165 100644
--- x/qtbase/src/gui/accessible/qaccessible_base.h
+++ y/qtbase/src/gui/accessible/qaccessible_base.h
@@ -266,6 +266,7 @@ public:
         WebDocument    = 0x00000084,
         Section        = 0x00000085,
         Notification   = 0x00000086,
+        Switch         = 0x00000087,
 
         // IAccessible2 roles
         // IA2_ROLE_CANVAS = 0x401, // An object that can be drawn into and to manage events from the objects drawn into it
diff --git x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
index 8de9758bb0f7a1f7c27c3cd1081a2074f8d10738..4222235d05a613093dded9c63ee96a2628a82494 100644
--- x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
@@ -475,7 +475,6 @@ namespace QtAndroidAccessibility
         case QAccessible::Role::Link:
         {
             if (state.checkable)
-                // There is also a android.widget.Switch for which we have no match.
                 return QStringLiteral("android.widget.ToggleButton");
             return QStringLiteral("android.widget.Button");
         }
@@ -483,6 +482,8 @@ namespace QtAndroidAccessibility
             // As of android/accessibility/utils/Role.java::getRole a CheckBox
             // is NOT android.widget.CheckBox
             return QStringLiteral("android.widget.CompoundButton");
+        case QAccessible::Role::Switch:
+            return QStringLiteral("android.widget.Switch");
         case QAccessible::Role::Clock:
             return QStringLiteral("android.widget.TextClock");
         case QAccessible::Role::ComboBox:
diff --git x/qtbase/src/plugins/platforms/cocoa/qcocoaaccessibility.mm y/qtbase/src/plugins/platforms/cocoa/qcocoaaccessibility.mm
index d77035a776aa5dc7ac258a0147ffede4d6587920..6063795d6b4e9f9ee9d2ad0d431d84e49abd4b75 100644
--- x/qtbase/src/plugins/platforms/cocoa/qcocoaaccessibility.mm
+++ y/qtbase/src/plugins/platforms/cocoa/qcocoaaccessibility.mm
@@ -131,6 +131,7 @@ static void populateRoleMap()
     roleMap[QAccessible::ComboBox] = NSAccessibilityComboBoxRole;
     roleMap[QAccessible::RadioButton] = NSAccessibilityRadioButtonRole;
     roleMap[QAccessible::CheckBox] = NSAccessibilityCheckBoxRole;
+    roleMap[QAccessible::Switch] = NSAccessibilityCheckBoxRole;
     roleMap[QAccessible::StaticText] = NSAccessibilityStaticTextRole;
     roleMap[QAccessible::Table] = NSAccessibilityTableRole;
     roleMap[QAccessible::StatusBar] = NSAccessibilityStaticTextRole;
@@ -203,6 +204,8 @@ NSString *macSubrole(QAccessibleInterface *interface)
         return NSAccessibilitySecureTextFieldSubrole;
     if (interface->role() == QAccessible::PageTab)
         return NSAccessibilityTabButtonSubrole;
+    if (interface->role() == QAccessible::Switch)
+        return NSAccessibilitySwitchSubrole;
     return nil;
 }
 
@@ -327,8 +330,11 @@ NSString *getTranslatedAction(const QString &qtAction)
 QString translateAction(NSString *nsAction, QAccessibleInterface *interface)
 {
     if ([nsAction compare: NSAccessibilityPressAction] == NSOrderedSame) {
-        if (interface->role() == QAccessible::CheckBox || interface->role() == QAccessible::RadioButton)
+        if (interface->role() == QAccessible::CheckBox
+            || interface->role() == QAccessible::RadioButton
+            || interface->role() == QAccessible::Switch) {
             return QAccessibleActionInterface::toggleAction();
+        }
         return QAccessibleActionInterface::pressAction();
     } else if ([nsAction compare: NSAccessibilityIncrementAction] == NSOrderedSame)
         return QAccessibleActionInterface::increaseAction();
diff --git x/qtbase/src/plugins/platforms/ios/quiaccessibilityelement.mm y/qtbase/src/plugins/platforms/ios/quiaccessibilityelement.mm
index 882c60dadab6d8c53c651d93f76d9f50eac3a958..d388f2ab33bbb83ee2502acb2a55f277d081a568 100644
--- x/qtbase/src/plugins/platforms/ios/quiaccessibilityelement.mm
+++ y/qtbase/src/plugins/platforms/ios/quiaccessibilityelement.mm
@@ -114,6 +114,9 @@ QT_NAMESPACE_ALIAS_OBJC_CLASS(QMacAccessibilityElement);
             || iface->role() == QAccessible::RadioButton)
             return @"";
 
+        if (iface->role() == QAccessible::Switch)
+            return state.checked ? @"1" : @"0";
+
         return state.checked
                 ? QCoreApplication::translate(ACCESSIBILITY_ELEMENT, AE_CHECKED).toNSString()
                 : QCoreApplication::translate(ACCESSIBILITY_ELEMENT, AE_UNCHECKED).toNSString();
@@ -168,6 +171,12 @@ QT_NAMESPACE_ALIAS_OBJC_CLASS(QMacAccessibilityElement);
                || accessibleRole == QAccessible::RadioButton) {
         if (state.checked)
             traits |= UIAccessibilityTraitSelected;
+    } else if (accessibleRole == QAccessible::Switch) {
+        if (@available(iOS 17.0, *)) {
+            traits |= UIAccessibilityTraitToggleButton;
+        } else {
+            traits |= UIAccessibilityTraitButton;
+        }
     } else if (accessibleRole == QAccessible::EditableText) {
         static auto defaultTextFieldTraits = []{
             auto *textField = [[[UITextField alloc] initWithFrame:CGRectZero] autorelease];
diff --git x/qtbase/src/plugins/platforms/wasm/qwasmaccessibility.cpp y/qtbase/src/plugins/platforms/wasm/qwasmaccessibility.cpp
index 2e430176bec2b124a1ff6ee01dbf7333d2ccc1ec..b51c4e100401efc8812691fb096ed5a7522400a7 100644
--- x/qtbase/src/plugins/platforms/wasm/qwasmaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/wasm/qwasmaccessibility.cpp
@@ -168,6 +168,17 @@ emscripten::val QWasmAccessibility::createHtmlElement(QAccessibleInterface *ifac
 
         } break;
 
+        case QAccessible::Switch: {
+            element = document.call<emscripten::val>("createElement", std::string("button"));
+            setAttribute(element, "type", "button");
+            setAttribute(element, "role", "switch");
+            if (iface->state().checked)
+                setAttribute(element, "aria-checked", "true");
+            else
+                setAttribute(element, "aria-checked", "false");
+            addEventListener(element, "change");
+        } break;
+
         case QAccessible::RadioButton: {
             element = document.call<emscripten::val>("createElement", std::string("input"));
             element.call<void>("setAttribute", std::string("type"), std::string("radio"));
@@ -483,6 +494,28 @@ void QWasmAccessibility::handleCheckBoxUpdate(QAccessibleEvent *event)
     break;
     }
 }
+
+void QWasmAccessibility::handleSwitchUpdate(QAccessibleEvent *event)
+{
+    switch (event->type()) {
+    case QAccessible::Focus:
+    case QAccessible::NameChanged: {
+        setHtmlElementTextName(event->accessibleInterface());
+    } break;
+    case QAccessible::StateChanged: {
+        QAccessibleInterface *accessible = event->accessibleInterface();
+        const emscripten::val element = getHtmlElement(accessible);
+        if (accessible->state().checked)
+            setAttribute(element, "aria-checked", "true");
+        else
+            setAttribute(element, "aria-checked", "false");
+    } break;
+    default:
+        qCDebug(lcQpaAccessibility) << "TODO: implement handleSwitchUpdate for event" << event->type();
+    break;
+    }
+}
+
 void QWasmAccessibility::handleToolUpdate(QAccessibleEvent *event)
 {
     QAccessibleInterface *iface = event->accessibleInterface();
@@ -729,6 +762,9 @@ void QWasmAccessibility::notifyAccessibilityUpdate(QAccessibleEvent *event)
     case QAccessible::CheckBox:
         handleCheckBoxUpdate(event);
     break;
+    case QAccessible::Switch:
+        handleSwitchUpdate(event);
+    break;
     case QAccessible::EditableText:
         handleLineEditUpdate(event);
     break;
diff --git x/qtbase/src/plugins/platforms/wasm/qwasmaccessibility.h y/qtbase/src/plugins/platforms/wasm/qwasmaccessibility.h
index c4be7f0d72a14143635d0f70d595498dc0bec3b0..71d2794f2eab6c97eee8aa2995a849f7e4f747c3 100644
--- x/qtbase/src/plugins/platforms/wasm/qwasmaccessibility.h
+++ y/qtbase/src/plugins/platforms/wasm/qwasmaccessibility.h
@@ -56,6 +56,7 @@ private:
     void handleStaticTextUpdate(QAccessibleEvent *event);
     void handleButtonUpdate(QAccessibleEvent *event);
     void handleCheckBoxUpdate(QAccessibleEvent *event);
+    void handleSwitchUpdate(QAccessibleEvent *event);
     void handleDialogUpdate(QAccessibleEvent *event);
     void handleMenuUpdate(QAccessibleEvent *event);
     void handleToolUpdate(QAccessibleEvent *event);
diff --git x/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp y/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
index 3417a3892115ddf303e612ab5bca62aa1d32dada..29a4a0b9651da06540a3e195b54f92ee1e7f6ab8 100644
--- x/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
+++ y/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
@@ -83,8 +83,9 @@ void QWindowsUiaMainProvider::notifyStateChange(QAccessibleStateChangeEvent *eve
 {
     if (QAccessibleInterface *accessible = event->accessibleInterface()) {
         if (event->changedStates().checked || event->changedStates().checkStateMixed) {
-           // Notifies states changes in checkboxes.
-           if (accessible->role() == QAccessible::CheckBox) {
+            // Notifies states changes in checkboxes and switches.
+            if (accessible->role() == QAccessible::CheckBox
+                || accessible->role() == QAccessible::Switch) {
                 if (auto provider = providerForAccessible(accessible)) {
                     long toggleState = ToggleState_Off;
                     if (accessible->state().checked)
diff --git x/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiautils.cpp y/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiautils.cpp
index 7536ece4dbc940a9dd171a1f46497e46c9388f9b..268961486e75ea82d26172309051056d7c61369c 100644
--- x/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiautils.cpp
+++ y/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiautils.cpp
@@ -130,6 +130,7 @@ long roleToControlTypeId(QAccessible::Role role)
         {QAccessible::EditableText, UIA_EditControlTypeId},
         {QAccessible::Button, UIA_ButtonControlTypeId},
         {QAccessible::CheckBox, UIA_CheckBoxControlTypeId},
+        {QAccessible::Switch, UIA_ButtonControlTypeId},
         {QAccessible::RadioButton, UIA_RadioButtonControlTypeId},
         {QAccessible::ComboBox, UIA_ComboBoxControlTypeId},
         {QAccessible::ProgressBar, UIA_ProgressBarControlTypeId},
